src/storage/src/hummock/state_store.rs

```diff
@@ -95,65 +93,61 @@ impl HummockStorage {
-                    let table_infos = &level.get_table_infos()[start_table_idx..=end_table_idx];
+            let table_infos =
+                prune_ssts(level.get_table_infos().iter(), &key_range, backward, None);
+            if table_infos.is_empty() {
+                continue;
+            }
 
-                    if backward {
-                        overlapped_backward_iters.push(Box::new(BackwardConcatIterator::new(
-                            table_infos.iter().rev().cloned().collect(),
-                            self.sstable_store(),
-                        ))
-                            as BoxedBackwardHummockIterator);
-                    } else {
-                        overlapped_forward_iters.push(Box::new(ConcatIterator::new(
-                            table_infos.iter().cloned().collect_vec(),
-                            self.sstable_store(),
-                        ))
-                            as BoxedForwardHummockIterator);
-                    };
-                }
+            for table_info in table_infos.into_iter().rev() {
+                let table = self.sstable_store.sstable(table_info.id).await?;
+                if backward {
+                    overlapped_backward_iters.push(Box::new(BackwardSSTableIterator::new(
+                        table,
+                        self.sstable_store(),
+                    ))
+                        as BoxedBackwardHummockIterator);
+                } else {
+                    overlapped_forward_iters.push(Box::new(SSTableIterator::new(
+                        table,
+                        self.sstable_store(),
+                    ))
+                        as BoxedForwardHummockIterator);
+                };
             }
+            // LevelType::Nonoverlapping => {
+            // if level.get_table_infos().is_empty() {
+            // continue;
+            // }
+            //
+            // let start_table_idx = match key_range.start_bound() {
+            // Included(key) | Excluded(key) => search_sst_idx(level, key),
+            // _ => 0,
+            // };
+            // let end_table_idx = match key_range.end_bound() {
+            // Included(key) | Excluded(key) => search_sst_idx(level, key),
+            // _ => level.table_infos.len().saturating_sub(1),
+            // };
+            // assert!(
+            // start_table_idx < level.table_infos.len()
+            // && end_table_idx < level.table_infos.len()
+            // );
+            // let table_infos = &level.get_table_infos()[start_table_idx..=end_table_idx];
+            //
+            // if backward {
+            // overlapped_backward_iters.push(Box::new(BackwardConcatIterator::new(
+            // table_infos.iter().rev().cloned().collect(),
+            // self.sstable_store(),
+            // ))
+            // as BoxedBackwardHummockIterator);
+            // } else {
+            // overlapped_forward_iters.push(Box::new(ConcatIterator::new(
```

Comment:
After this PR, we treat all SSTs as overlapping SSTs but read performance on SSTs generated by range compaction should not be affected after we introduce hash compaction. One thing we can do here is to check whether SSTs after `prune_ssts` are non-overlapping and keep using ConcatIterator if that is the case.

Reply:
fixed